accepted/wasmtime-one-dot-oh.md

```diff
@@ -0,0 +1,558 @@
+which could result in some releases being a slightly different width apart.
+
+The goal is to publish work and improvements in Wasmtime on a relatively rapid
+schedule to ensure that the latest-and-greatest is available for usage. This is
+also intended to signal to users that consumers of Wasmtime should expect to
+stay up-to-date with Wasmtime and it's not appropriate to ossify a version for
+years and expect it to continue to work and receive everlasting maintenance.
+
+Each release of Wasmtime will bump the major version number of Wasmtime itself.
+This means, for example, that Wasmtime 2.0 will be released one month after
+Wasmtime 1.0. After one year Wasmtime will be at 12.0. At this time it is not
+planned that "minor" releases will be made of the 1.1.0 variety. Note that this
+means means that not all major releases will actually contain breaking changes.
+The reason we propose to nevertheless always bump the major version is that this
+allows us to keep Wasmtime's version number in sync with that of its various
+language-specific embeddings: as a developer using Wasmtime, you shouldn't have
+to worry about how how the version number of your language's Wasmtime embedding
+lines up with that of Wasmtime itself. E.g., if you use`wasmtime-py 7.0`, you
+can be sure that you're using Wasmtime 7.0. Trying to keep major version bumps
+to a minimum while keeping version numbers aligned would force us to still bump
+Wasmtime's and all embeddings' version numbers whenever there's a breaking
+change in even a single language-specific embedding. As noted in the previous
+section, though, each release is expected to be a relatively hassle-free
+upgrade, so while API-breaking changes are allowed they're not necessarily
+encouraged.
+
+Wasmtime will continue to create a tag for all released versions of Wasmtime
+with a corresponding GitHub release as is done today for the CLI and C API
+binaries. Embeddings will all be tagged and released as appropriate to
+language-specific package managers (such as crates.io and PyPI). Wasmtime will
+always be released from the `main` branch of the Wasmtime repository itself,
+which means that all development is happening on `main` and once something lands
+it's guaranteed to be in the next release. Note that bug fixes and such for
+historical releases are discussed later in this proposal.
+
+Releasing a new version of Wasmtime every 4 weeks can be quite rapid for some
+users who don't necessarily want to stay up-to-date with the latest and greatest
+of Wasmtime, but still want the stability of a production-ready WebAssembly
+runtime. For these users, this leads well into the next section ...
+
+### Long-term Support Releases
+
+Wasmtime will support some releases for an extended period of time relative to
+other releases. These releases will be known as "long term support" releases or
+LTS releases. **Wasmtime will support two active LTS versions at any point in
```

Comment:
Maintaining LTS releases is a costly process, so starting with two when it's not clear that even one is necessary seems like an overkill. I'd suggest starting with one, perhaps with a slightly longer cycle (~6 months), and adding another one only if there is a clear demand for it.

Reply:
The worry we have about only having a single LTS release is that it basically means LTS users have a zero-length upgrade window when a new LTS release is published. That seems counter to what someone in need of an LTS release would want :)

Plus, our hope is that we'll be able to automate our processes around releases to a very high degree, and to make use of that same automation for LTS releases as well—which ideally will address some of the cost concerns.

At the very least, I'd hope that maintaining two LTS releases wouldn't be *that* much more burdensome than maintaining a single one, in particular given the tightly circumscribed scope of changes we intend to make available to LTS releases.

I think it's worth at least thinking through what it'd mean to start without an LTS release at all: you're right that the need for one is more derived from precedent in other projects than from immediate requests from Wasmtime users.

An option could be to split the LTS part out into a separate proposal that could simmer for a bit longer. The one worry I'd have about that is that it's probably easier to establish the required processes for all of this in one go, so getting up the energy to do the LTS stuff later might simply be harder. But @alexcrichton would much better be able to comment on that part than me :)

Reply:
I meant the cost of backporting fixes, which often means rewriting the change from scratch, since trivial cherry-picks to a branch from a few months ago rarely work in an actively developed codebase... Having said that, I'm not very familiar with the codebase or the development pace of Wasmtime, so perhaps that won't be an issue.

Reply:
AFAIK most of us working on Wasmtime haven't dealt much historically with maintaining LTS releases of software, so in that sense that may be why we're being a bit overly ambitiuos here. Given how different the LTS process is I don't forsee it being easier to add now vs later if we decide to do it, so in that sense I don't think there's urgency to add it earlier while the release process is all in someone's head.

An alternative model could be one where the LTS windows overlap for 1 release instead of 5 releases. For example 5.0 would be an LTS release for the 5.0-11.0 releases, but when 12.0 is released only 10.0 is a supported LTS release. Something like that means that our window for supporting multiple LTS releases is much smaller and because 11.0 is probably pretty similar to 10.0 any work for backports would primarily be from 11.0 to 5.0 rather than 11.0 to 10.0 (which is likely cherry-pick-able). This would still have the property where there's a window where users can upgrade but still get bug fixes if necessary.

Naturally though I'd expect that anything here can change at any time with sufficient consensus. It's probably a lot easier to add an LTS process than it is to remove one as well. Given that we don't have anyone specifically asking for LTS it may not be a bad idea to not support it initially. We could have some initial releases and see if users find the update-every-4-week cadence to be too onerous. We'd probably, in that case, update the bugfix policy to backport to the current release at the time of the bugfix and perhaps the last 1-2 releases prior to that as well.

Overall I don't personally know the best thing for Wasmtime at this time. We made an initial guess with 2 LTS releases at this cadence, but it's mostly a guess that it'll work out.

Reply:
If customers aren't requesting an LTS, it might be worth postponing until the release process has had some time to mature. I completely agree that backporting is very costly, even in relatively mature code bases, so it could be too much work for little gain.

And sorry if I missed it, but is there a plan for stabilization and testing before a release drops, or are we going for an "always releasable trunk"? Particularly if we go for LTS releases, it would be really useful to give developers and users a reasonable amount of time to test before hand, otherwise we'd probably just get a bunch of issues a couple of weeks after the release and it wouldn't be great for trust to push out a buggy LTS.

Reply:
I posted a comment at the bottom about this, but personally I agree with y'all's warnings and conclusions that maybe we don't need LTS at all.

For releases yeah the plan is that `main` is always green with CI and always release-able.